File: Roadmap/09 - HERENCIA/typescript/AChapeton.ts
class Animal{
  constructor(name: string, size: string, isDomestic: Boolean){
    this.name = name
    this.size = size
    this.isDomestic = isDomestic
  }
}

class Dog extends Animal{
  get makeDogSound(){
    return this.dogSound()
  }
  
  dogSound() {
    return 'bark' + ' ' + this.name
  }
}

class Cat extends Animal{
  get makeCatSound(){
    return this.catSound()
  }
  
  catSound() {
    return 'meow' + ' ' + this.name
  }
}

const myDog = new Dog('Dolly', 'small', true)
console.log(myDog.makeDogSound)

const myCat = new Cat('Snowball', 'small', true)
console.log(myCat.makeCatSound)




/* DIFICULTAD EXTRA */

class Empleado{
  constructor(id: string, employeeName: string){
    this.id = id
    this.employeeName = employeeName
  }

  get makePresentation(){
    return this.presentation()
  }

  presentation(){
    return 'Hola, mi nombre es ' + this.employeeName
  }
}

class Gerente extends Empleado{
  constructor(id: string, employeeName: string, clients: Array<string> = []){
  super(id, employeeName)
    this.clients = clients
  }

  get showClientsList(){
    return this.clientsList()
  }

  clientsList(){
    return this.clients
  }

  addNewClient(companyName: string){
    this.clients.push(companyName)
  }

}

const newGerente = new Gerente('1', 'Andres')
console.log(newGerente.makePresentation)
console.log(newGerente.showClientsList)
newGerente.addNewClient('ITO')
console.log(newGerente.showClientsList)


class GerenteProyecto extends Empleado{
  constructor(id: string, employeeName: string, projects: Array<string> = []){
    super(id, employeeName)
    this.projects = projects
  }

  get showProjectsList(){
    return this.projectsList()
  }

  projectsList(){
    return this.projects
  }

  startProject(projectName: string){
    this.projects.push(projectName)
  }

  finishProject(projectName: string){
    this.projects = this.projects.filter(project => project !== projectName)
  }  
}


const newGerenteProyecto = new GerenteProyecto('2', 'Edith')
console.log(newGerenteProyecto.makePresentation)
console.log(newGerenteProyecto.showProjectsList)
newGerenteProyecto.startProject('Web page')
newGerenteProyecto.startProject('Mobile app')
newGerenteProyecto.startProject('Logo refresh')
console.log(newGerenteProyecto.showProjectsList)
newGerenteProyecto.finishProject('Mobile app')
console.log(newGerenteProyecto.showProjectsList)


class Programador extends Empleado{
  constructor(id: string, employeeName: string, tasks: Array<string> = []){
    super(id, employeeName)
    this.tasks = tasks
  }

  get showTasksList(){
    return this.tasksList()
  }

  tasksList(){
    return this.tasks
  }

  startNewTask(newTask: string){
    this.tasks.push(newTask)
  }

  finishTask(taskName: string){
    this.tasks = this.tasks.filter(task => task !== taskName)
  }
}

const newProgramador = new Programador('3', 'Jesus')
console.log(newProgramador.makePresentation)
console.log(newProgramador.showTasksList)
newProgramador.startNewTask('refactor sign in function')
newProgramador.startNewTask('fix form styles')
newProgramador.startNewTask('create props TS interface')
console.log(newProgramador.showTasksList)
newProgramador.finishTask('fix form styles')
console.log(newProgramador.showTasksList)

File: Roadmap/09 - HERENCIA/python/AChapeton.py
class Animal:
  def __init__(self, name, sound):
    self.name = name
    self.sound = sound
  
  def makeSound(self):
    print(f"{self.name} dice {self.sound}")

class Dog(Animal):
  def __init__(self, name):
    super().__init__(name, 'Woof')

class Cat(Animal):
  def __init__(self, name):
    super().__init__(name, "Meow")


myDog = Dog("Dolly")
myDog.makeSound()

myCat = Cat("Snowball")
myCat.makeSound()


# DIFICULTAD EXTRA

class Empleado:
  def __init__(self, id, name, tasks, staff):
    self.id = id
    self.name = name
    self.tasks = tasks
    self.staff = staff
  
  def makePresentation(self):
    print(f"Hola, mi nombre es {self.name}")

  def showTasksList(self):
    print(f"Lista de tareas: {self.tasks}")

  def showTeam(self):
    print(f"A cargo de: {self.staff}")

class Gerente(Empleado):
  def __init__(self, name):
    super().__init__('Gerente', name, {'Obtener clientes', 'Ventas'}, {'Gerente de Proyecto', 'Programador'})

class GerenteDeProyecto(Empleado):
  def __init__(self, name):
    super().__init__('Gerente de Proyecto', name, {'Gestionar proyectos', 'Crear tickets', 'Agendar meets'}, {'Programador'})

class Programador(Empleado):
  def __init__(self, name):
    super().__init__('Programador', name, {'Resolver tickets'}, {'No tiene nadie a cargo'})

gerente = Gerente("Brais Moure")
gerente.makePresentation()
gerente.showTasksList()
gerente.showTeam()

gerenteDeProyecto = GerenteDeProyecto("Edith Romero")
gerenteDeProyecto.makePresentation()
gerenteDeProyecto.showTasksList()
gerenteDeProyecto.showTeam()

programador = Programador("Andres Chapeton")
programador.makePresentation()
programador.showTasksList()
programador.showTeam()

